src/main/java/org/skriptlang/skript/scheduler/AsyncTask.java

```diff
@@ -0,0 +1,185 @@
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with Skript.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Copyright Peter Güttinger, SkriptLang team and contributors
+ */
+package org.skriptlang.skript.scheduler;
+
+import java.util.concurrent.TimeUnit;
+
+import org.bukkit.plugin.Plugin;
+
+import ch.njol.skript.Skript;
+import ch.njol.util.Closeable;
+
+/**
+ * Represents a Task that calls its runnable asynchronously from the main server thread.
+ */
+public abstract class AsyncTask implements Runnable, Closeable {
+
+	private final TimeUnit unit;
+	private final Plugin plugin;
+	private long period = -1;
+
+	public AsyncTask(long delayInTicks, long periodInTicks) {
+		this(Skript.getInstance(), delayInTicks, periodInTicks);
+	}
+
+	public AsyncTask(Plugin plugin, long delayInTicks, long periodInTicks) {
+		this(plugin, (delayInTicks / 20) * 1000, (periodInTicks / 20) * 1000, TimeUnit.MILLISECONDS);
+	}
+
+	public AsyncTask(long delay, long period, TimeUnit unit) {
+		this(Skript.getInstance(), period, unit);
+	}
+
+	public AsyncTask(Plugin plugin, long delay, long period, TimeUnit unit) {
+		this.plugin = plugin;
+		this.period = period;
+		this.unit = unit;
+		TaskManager.run(this, delay, unit);
+	}
+
+	public AsyncTask(long delayInTicks) {
+		this(Skript.getInstance(), delayInTicks);
+	}
+
+	public AsyncTask(Plugin plugin, long delayInTicks) {
+		this(plugin, (delayInTicks / 20) * 1000, TimeUnit.MILLISECONDS);
+	}
+
+	public AsyncTask(long delay, TimeUnit unit) {
+		this(Skript.getInstance(), delay, unit);
+	}
+
+	public AsyncTask(Plugin plugin, long delay, TimeUnit unit) {
+		this.plugin = plugin;
+		this.unit = unit;
+		TaskManager.run(this, delay, unit);
+	}
+
+	/**
+	 * @return The plugin that is calling this task.
+	 */
+	public Plugin getPlugin() {
+		return plugin;
+	}
+
+	/**
+	 * @return The period of this task if it's repeating, otherwise -1
+	 */
+	public long getPeriod() {
+		return period;
+	}
+
+	/**
+	 * @return The amount the delay and period represent in TimeUnit.
+	 */
+	public TimeUnit getTimeUnit() {
+		return unit;
+	}
+
+	/**
+	 * @return Whether this task is still running, i.e. whether it will run later or is currently running.
+	 */
+	public final boolean isAlive() {
+		return TaskManager.isAlive(this);
+	}
+
+	/**
+	 * Cancels this task.
+	 * @return true if the task was successfully cancelled.
+	 */
+	public final boolean cancel() {
+		return TaskManager.cancel(this);
+	}
+
+	@Override
+	public void close() {
+		cancel();
+	}
+
+	/**
+	 * Re-schedules the task to run next after the given delay. If this task was repeating it will continue so using the same period as before.
+	 * 
+	 * @param delay The delay using the existing TimeUnit.
+	 */
+	public void setNextExecution(long delay) {
+		setNextExecution(delay, unit);
+	}
+
+	/**
+	 * Re-schedules the task to run next after the given delay. If this task was repeating it will continue so using the same period as before.
```

Comment:
Does this line mean that this method has no effect on repeating tasks? I think it should be rewritten to be a little clearer